Source/JavaScriptCore/runtime/RegExpObjectInlines.h

```diff
@@ -225,4 +226,101 @@ JSValue collectMatches(VM& vm, JSGlobalObject* globalObject, JSString* string, c
     return array;
 }
 
+struct GlobalAtomMeta {
+    bool isSubstring() { return !!base; }
+
+    JSString* base { nullptr };
+    unsigned offset { 0 };
+    unsigned length { 0 };
+
+    size_t numberOfMatches { 0 };
+    size_t startIndex { 0 };
+};
+
+template<typename Char>
+inline void oneCharMatches(std::span<const Char> input, Char pattern, GlobalAtomMeta& meta)
+{
+    for (size_t i = meta.startIndex; i < input.size(); ++i) {
+        if (input[i] != pattern)
+            continue;
+        meta.numberOfMatches++;
+        meta.startIndex = i + 1;
+        if (meta.numberOfMatches > MAX_STORAGE_VECTOR_LENGTH)
+            return;
+    }
+}
+
+inline void genericMatches(StringView input, StringView pattern, GlobalAtomMeta& meta)
+{
+    size_t found = input.find(pattern, meta.startIndex);
+    while (found != notFound) {
+        meta.startIndex = found + pattern.length();
+        meta.numberOfMatches++;
+        found = input.find(pattern, meta.startIndex);
+        if (meta.numberOfMatches > MAX_STORAGE_VECTOR_LENGTH)
+            return;
+    }
+}
+
+inline JSValue collectGlobalAtomMatches(JSGlobalObject* globalObject, JSString* string, RegExp* regExp)
+{
+    VM& vm = globalObject->vm();
+    auto scope = DECLARE_THROW_SCOPE(vm);
+
+    // Get the cache.
+    auto& cache = globalObject->regExpGlobalData().substringGlobalAtomCache();
+    GlobalAtomMeta meta;
+    cache.tryGet(string, regExp, meta);
+
+    // Keep the substring info above since the following will resolve the substring to a non-rope.
+    auto input = string->value(globalObject);
+    RETURN_IF_EXCEPTION(scope, { });
+
+    // Find the matches.
+    const String& pattern = regExp->atom();
+    ASSERT(!pattern.isEmpty() && !pattern.isNull());
+    if (pattern.is8Bit()) {
+        if (input->is8Bit()) {
+            if (pattern.length() == 1)
+                oneCharMatches(input->span8(), pattern.span8()[0], meta);
+            else
+                genericMatches(input, pattern, meta);
+        } else {
+            if (pattern.length() == 1)
+                oneCharMatches(input->span16(), pattern.characterAt(0), meta);
+            else
+                genericMatches(input, pattern, meta);
+        }
+    } else {
+        if (input->is8Bit())
+            genericMatches(input, pattern, meta);
+        else {
+            if (pattern.length() == 1)
+                oneCharMatches(input->span16(), pattern.characterAt(0), meta);
+            else
+                genericMatches(input, pattern, meta);
+        }
+    }
+
+    if (!meta.numberOfMatches)
+        return jsNull();
+    if (meta.numberOfMatches > MAX_STORAGE_VECTOR_LENGTH) {
+        throwOutOfMemoryError(globalObject, scope);
+        return jsUndefined();
+    }
+
+    // Construct the array
+    JSArray* array = constructEmptyArray(globalObject, nullptr, meta.numberOfMatches);
```

Comment:
Let's consider doing super fast allocation here for major case.
1. First, get the pattern string
2. Next, try allocate Butterfly with the size (if it is not possible (due to slow array structure etc., use the current slow code as a fallback)
3. Fill the butterfly with the (1)'s pointer. You can use memset_pattern8 on Darwin further
4. Create array with the butterfly.